### PR TITLE
Web [Charts]: Fix wrong rendering order and focus/hover line state

### PR DESCRIPTION
**Pull request by vovakulikov**
Fixes: https://github.com/sourcegraph/sourcegraph/issues/38427
Fixes: https://github.com/sourcegraph/sourcegraph/issues/39789

# Background

Imagine that you have a line chart with two lines on it, with five dots in each line. And lines are located close enough to each other, so close that it is usually impossible to aim at or see any of lines points behind other lines.

Untitled-2022-08-06-1827-2

So we have to change the z-index somehow for SVG elements to pull hovered/focused line to the front. According to the SVG spec, we don't have any control over elements order except the actual DOM elements order.
>[3.3 Rendering Order](https://www.w3.org/TR/SVG11/render.html#RenderingOrder) Elements in an SVG document fragment have an implicit drawing order, with the first element in the SVG document fragment getting "painted" first. Subsequent elements are painted on top of previously painted elements.

Based on this, we need to change SVG DOM elements' order when we hover points on the chart. But if we're changing elements order on some trigger (mouse hover or keyboard focus navigation), we're breaking the normal focus management for the chart. 

Let's imagine you're focusing on the first point on the chart. We immediately change the order in a way to bring this line on top (this means this line now is the last one in the chart according to the DOM hierarchy). This means that when you end focus this line, the focus naturally will go away from the chart since you're going from the last element on the chart already. 

## Solution 
Instead of changing the order of chart lines (bars, arcs, whatever), we will render the line copy as the last element on the chart (so visually, it will be over all other elements on the chart) and mark this as hidden for keyboard and all a11y technologies). In this case, we don't break anything about focus management for the SVG elements. 

```html
<svg>
  <line .../>
  <!-- Focus is somewhere in the second line -->
  <line .../> 
  <line .../>

   <!-- Copy of the second line element that it's rendered as the last element  -->
  <line tabIndex="-1" pointer-events="none" aria-hidden="hidden".../> 
</svg>
```

## Test plan
- Check that focus works properly for line charts with more than one series (lines)
- Check that multi lines charts on the code insight dashboard page (or analytics pages) don't have a weird z-index problem (on the initial mount and when you're hovering/focusing different lines on the chart)

## App preview:

- [Web](https://sg-web-vk-old-chart-line-fix.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-yukvjjrrzj.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by felixfbecker** [comment]
> Imagine that you have a line chart with two lines on it, with five dots in each line. And they are located close enough to each other, so close that it is usually impossible to aim at any of the points.

> So we have to change the z-index somehow for SVG elements. According to the SVG spec, we don't have any control over elements order except the actual DOM elements order.

I can't really follow the reasoning here – how does the first statement imply the second statement? Don't we have hover logic so that it always uses the "closest" line, so in the example given you could put your cursor either below or above the line to "hover" it? What does that have to do with z indexes?

The linked issues also don't imply to me that we need to dynamically change the z index of any lines, just that we need to render without transparency when one is hovered (#38427) and that there is some kind of bug with mismatching z-indexes between points and the line they belong to (#39789).

It seems that if anything, adding logic to dynamically change z-indexes would introduce complexity and make z-indexing bugs more likely?

**Comment by vovakulikov** [comment]
@felixfbecker 

I think I should update this PR description if it's not clear why we had this dynamically z-index logic. 
Just for the record I write this as a separate message here. (and if it's clear I will update the PR)

**TL;DR**

> It seems that if anything, adding logic to dynamically change z-indexes would introduce complexity and make z-indexing bugs more likely?

This is exactly right, in this PR I render lines and their points together and I don't change their order in SVG at all. Visually problems with pulling active lines over other lines we solve with fake (ghost) line element. This line is hidden from voice over or keyboard navigation and it's rendered as the last element in SVG so it's always placed over other lines.

What about https://github.com/sourcegraph/sourcegraph/issues/38427? This issue was fixed during small refactoring of line components and it's just a bonus fix here. 

**A bit history of the line chart below**

So based on our designs we always should pull any active, hovered or focus line on top of other lines (to make active line more visible and visually accessible). The only way we came up to achieve this was changing order of line elements (because SVG doesn't have z-index). This was a bad decision because as I described in this PR description it messes up with focus flow really bad. 

Next solution after that was rendering points separately from lines elements and changing only line elements order (points order remains the same). It allowed us to change order of lines to solve this problem with z-index and pulling active line over other lines and in the same time don't change order of points elements (which are focusable) and hense don't mess up with focus flow. 

But this solution actually also had problems (which I hadn't seen in first two times 🤦 ) because active line points won't be pulled over other points (cause their order isn't changing). It leads to sync ordering problem when lines and points orders are not sync (this issue https://github.com/sourcegraph/sourcegraph/issues/39789 is exactly about it). 

@felixfbecker does it make sense?

**Comment by felixfbecker** [comment]
> So based on our designs, we should always pull any active, hovered or focus line on top of other lines (to make the active line more visible and visually accessible).

@AlicjaSuska can we just drop this requirement? It seems like a small visual nice-to-have that clearly introduces a ton of complexity (z-indexes, focus management, accessibility tree order, various related bugs, ...). Even if we don't raise it to the front, we still enlarge the currently hovered point meaning it will stick out, and there is still the tooltip.

Curious in which design this was introduced? (since we didn't always have this)

**Comment by vovakulikov** [comment]
> It seems like a small visual nice-to-have that clearly introduces a ton of complexity (z-indexes, focus management, accessibility tree order, various related bugs, ...

Just to be clear (it wasn't clear from my message above), in this PR, we remove most of the complexity, we don't do anything with z-indexes anymore, focus management relies on DOM elements order, and we don't interrupt native focus in any way, accessibility tree order stays the same since we handle visual element and hide it from a11y tree.

@felixfbecker I disagree that this is just "a small visual nice-to-have" feature. Considering the usually relatively small size of our charts, this feature helps me a lot when I explore insights on the dashboard page.

File: client/web/src/charts/components/line-chart/LineChart.tsx [hunk 1]
```diff
@@ -50,6 +39,8 @@ export interface LineChartProps<Datum> extends SeriesLikeChart<Datum>, SVGProps<
     /** Whether to start Y axis at zero */
     zeroYAxisMin?: boolean
 
+    activeSeriesId?: string
```
Comment (vovakulikov) [inline client/web/src/charts/components/line-chart/LineChart.tsx:42]: Add this prop to support a controlled case when a consumer wants to bring a line to the front. This is used in the code insights charts

File: client/web/src/charts/components/line-chart/LineChart.tsx [hunk 2]
```diff
@@ -190,16 +170,9 @@ export function LineChart<D>(props: LineChartProps<D>): ReactElement | null {
...
-    const sortedSeries = useMemo(
-        () =>
-            [...activeSeries]
-                // resorts array based on hover state
-                // this is to make sure the hovered series is always rendered on top
-                // since SVGs do not support z-index, we have to render the hovered
-                // series last
-                .sort(series => sortByDataKey(series.id, activeSeriesId)),
-        [activeSeries, activeSeriesId]
+    const currentSeries = useMemo(
```
Comment (vovakulikov) [inline client/web/src/charts/components/line-chart/LineChart.tsx:173]: Find the line (series) by active points. We will use this later in the render call to render a copy of hovered/focus line at the very end of SVG

File: client/web/src/charts/components/line-chart/LineChart.tsx [hunk 3]
```diff
@@ -233,58 +206,47 @@ export function LineChart<D>(props: LineChartProps<D>): ReactElement | null {
...
-                        y={data => yScale(getDatumValue(data))}
-                        stroke={line.color}
-                        strokeLinecap="round"
-                        strokeWidth={2}
-                    />
-                ))}
-
                 {activeSeries.map(line => (
-                    <Group
+                    <LineDataSeries
```
Comment (vovakulikov) [inline client/web/src/charts/components/line-chart/LineChart.tsx:210]: Before this PR, we rendered lines and points separately now lines and points are rendered line by line together

File: client/web/src/charts/components/line-chart/utils/data-series-processing/helpers.ts [hunk 4]
```diff
@@ -2,40 +2,6 @@ import { isValidNumber } from '../data-guards'
 
 import { isStandardSeriesDatum, SeriesDatum } from './types'
 
-/**
```
Comment (vovakulikov) [inline client/web/src/charts/components/line-chart/utils/data-series-processing/helpers.ts:5]: I removed this filter helper since this filter isn't used anymore. And we no longer have cases when we need to filter null values

**Review by chwarwick — APPROVED** [review]
This change makes sense to me.  My only concern would be if the "duplicate" line would cause any new issues since that would now the item that is catching events that appears covered.  I tested it in the preview and had the experience that I expected when there were overlapping lines.

File: client/web/src/charts/components/line-chart/utils/data-series-processing/get-stacked-series-data.ts [hunk 5]
```diff
@@ -106,12 +91,12 @@ function findPreviousValueOnStack<Datum>(
...
-
-        if (hasExactMatch && stackedExactDatum.y1 !== null) {
-            return stackedExactDatum.y1
-        }
+        // const stackedExactDatum = stackedSeries.data[currentIndex]
+        // const hasExactMatch = +stackedExactDatum.x === +wantedDate
+        //
+        // if (hasExactMatch && stackedExactDatum.y1 !== null) {
+        //     return stackedExactDatum.y1
+        // }
```
Comment (felixfbecker) [inline client/web/src/charts/components/line-chart/utils/data-series-processing/get-stacked-series-data.ts:99]: Why is this code commented out?

Comment (vovakulikov) [reply client/web/src/charts/components/line-chart/utils/data-series-processing/get-stacked-series-data.ts]: Thanks for catching this. I was exploring stacked data charts and left this comment out by accident.
